File: 015_Three_Sum/015_1.py
from typing import List

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        length = len(nums)
        if length < 3 or not nums:
            return []
        nums.sort();
        for i in range(length - 2):
            if (nums[i] > 0):
                return res
            if i == 0 or nums[i] > nums[i - 1]:
                j = i + 1
                k = length - 1
                while j < k:
                    threeSum = nums[i] + nums[j] + nums[k]
                    if threeSum == 0:
                        res.append([nums[i], nums[j], nums[k]])
                        while j < k and nums[j] == nums[j + 1]:
                            j += 1
                        while j < k and nums[k] == nums[k - 1]:
                            k -= 1
                        j += 1
                        k -= 1
                    elif threeSum > 0:
                        k -= 1
                    else:
                        j += 1
        return res
                    

if __name__ == "__main__":
    nums = [-1,0,1,2,-1,-4]
    s = Solution()
    for r in s.threeSum(nums):
        print(r)

File: 015_Three_Sum/015_1.java
import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;

class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
    	List<List<Integer>> res = new ArrayList<>();
    	Arrays.sort(nums);
    	int length = nums.length;
    	for (int i = 0; i < length - 2; ++i) {
    		if (i == 0 || (i > 0 && nums[i - 1] < nums[i])) {
    			int j = i + 1;
    			int k = length - 1;
    			while (j < k) {
	    			int threeSum = nums[i] + nums[j] + nums[k];
	    			if (threeSum == 0) {
	    				res.add(Arrays.asList(nums[i], nums[j], nums[k]));
	    				while (j < k && nums[j] == nums[j + 1]) j++;
	    				while (j < k && nums[k] == nums[k - 1]) k--;
	    				j++;
	    				k--;
	    			} else if (threeSum < 0) {
	    				j++;
	    			} else {
	    				k--;
	    			}
    			}
    		}
    	}
    	return res;
    }
}


public class leetcode {
	public static void main(String[] args) {
		int[] nums = {-1,0,1,2,-1,-4};
		Solution s = new Solution();
		List<List<Integer>> res = s.threeSum(nums);
		for (List<Integer> a : res) {
			System.out.print('[');
			for (Integer b : a)
				System.out.print(b + " ");
			System.out.print(']');
			System.out.println();
		}
	}
}